sqlite_query.c

```diff
@@ -21,63 +21,82 @@
 #include "utils/builtins.h"
 #include "utils/lsyscache.h"
 
+#include "nodes/makefuncs.h"
+#include "catalog/pg_type.h"
+#include "parser/parse_type.h"
+
 static int32
-			sqlite_from_pgtyp(Oid pgtyp);
+			sqlite_affinity_eqv_to_pgtype(Oid pgtyp);
+static const char*
+            sqlite_datatype(int t);
 
 /*
  * convert_sqlite_to_pg: Convert Sqlite data into PostgreSQL's compatible data types
  */
-Datum
-sqlite_convert_to_pg(Oid pgtyp, int pgtypmod, sqlite3_stmt * stmt, int stmt_colid, AttInMetadata *attinmeta, AttrNumber attnum)
+NullableDatum
+sqlite_convert_to_pg(Oid pgtyp, int pgtypmod, sqlite3_stmt * stmt, int stmt_colid, AttInMetadata *attinmeta, AttrNumber attnum, int sqlite_value_affinity, int AffinityBehaviourFlags)
 {
 	Datum		value_datum = 0;
 	char	   *valstr = NULL;
-	int			col_type = sqlite3_column_type(stmt, stmt_colid);
-	int			sqlite_type = sqlite_from_pgtyp(pgtyp);
+	int			affinity_for_pg_column = sqlite_affinity_eqv_to_pgtype(pgtyp);
+	int         value_byte_size_blob_or_utf8 = sqlite3_column_bytes(stmt, stmt_colid); // Compute always, void text and void BLOB will be special cases
 
-	if (sqlite_type != col_type && col_type == SQLITE3_TEXT)
-		elog(ERROR, "invalid input syntax for type =%d, column type =%d", sqlite_type, col_type);
+	if (affinity_for_pg_column != sqlite_value_affinity && sqlite_value_affinity == SQLITE3_TEXT)
+	{
+	    /* Here will be human readable error message
+        const char    *sqlite_affinity = 0;
+    	const char    *pg_eqv_affinity = 0;
+	    const char    *pg_dataTypeName = 0;
+	    
+	    pg_dataTypeName = TypeNameToString(makeTypeNameFromOid(pgtyp, pgtypmod));
+		sqlite_affinity = sqlite_datatype(sqlite_value_affinity);
+		pg_eqv_affinity = sqlite_datatype(affinity_for_pg_column);
+		
+		value_datum = CStringGetDatum((char*) sqlite3_column_text(stmt, attnum));
+		elog(ERROR, "SQLite data affinity = \"%s\" disallowed for PostgreSQL type \"%s\" = SQLite \"%s\", value ='%s'", sqlite_affinity, pg_dataTypeName, pg_eqv_affinity, (char*)value_datum);
+	    */
+		elog(ERROR, "invalid input syntax for type =%d, column type =%d", affinity_for_pg_column, sqlite_value_affinity);
+	}
 
 	switch (pgtyp)
 	{
 		case BYTEAOID:
 			{
-				int			blobsize = sqlite3_column_bytes(stmt, stmt_colid);
-
-				value_datum = (Datum) palloc0(blobsize + VARHDRSZ);
-				memcpy(VARDATA(value_datum), sqlite3_column_blob(stmt, stmt_colid), blobsize);
-				SET_VARSIZE(value_datum, blobsize + VARHDRSZ);
-				return PointerGetDatum(value_datum);
+				// int			value_byte_size_blob_or_utf = sqlite3_column_bytes(stmt, stmt_colid); // Calculated always for detectind void values
+				value_datum = (Datum) palloc0(value_byte_size_blob_or_utf8 + VARHDRSZ);
```

Comment:
`value_byte_size_blob_or_utf8` is used only in this block, could you explain why we need define it outside?  
I do not understand your comment as well: `// Compute always, void text and void BLOB will be special cases`

Reply:
In our  [data transformation table](https://github.com/mkgrgis/sqlite_fdw/tree/readme-fix-and-additions#datatypes) empty value with `text` affinity is special case. For example for PostgreSQL `uuid` we *should* always threat void `text` value as `NULL`. For PostgreSQL `int` we can threat this value as `NULL` if a special boolean option is `true`. So for all PostgreSQL data types `if (value_byte_size_blob_or_utf8)` will be necessary.

Reply:
Thank for your explanation.   
I understood.